lwip_demo/lwip_demo.c

```diff
@@ -0,0 +1,333 @@
+   } else {
+      printf("eth_mac_irq: RDFO was empty.\n");
+      return;
+   }
+}
+
+/**
+ * netif_output: output handler/driver for FPGA (communicate with AXI Stream
+ * FIFO
+ * - update statistic counters
+ * - Read the buffer usage with TDFV
+ * - enter critical section
+ * - write 2 Bytes-wise into the AXI Stream FIFO Buffer
+ * - write the length of the written Bytes into TLR
+ * - Reset the ISR
+ * - exit critical section
+ */
+static err_t netif_output(struct netif *netif, struct pbuf *p)
+{
+   LINK_STATS_INC(link.xmit);MIB2_STATS_NETIF_ADD(netif, ifinoctets, p->tot_len);
+   int unicast = ((((uint16_t *) p->payload)[0] & 0x01) == 0);
+   if (unicast) {
+      MIB2_STATS_NETIF_INC(netif, ifinucastpkts);
+   } else {
+      MIB2_STATS_NETIF_INC(netif, ifinnucastpkts);
+   }
+
+   printf("netif_output: Writing to Stream FIFO and start transmission.\n");
+   uint32_t TDFV_before = *TDFV;
+   uint32_t restore_2 = or1k_critical_begin();
+   *TDR = (uint32_t) 0x00000002;
+   uint32_t left, tmp_len;
+   uint32_t buf_p = 0x0;
+   for (left = 0; left < ((p->tot_len) / 2); left = left + 2) {
+      buf_p = ((uint16_t *) p->payload)[left];
+      buf_p = buf_p << 16;
+      buf_p = buf_p | ((uint16_t *) p->payload)[left + 1];
+      *TDFD = swap_uint32(buf_p);
+   }
+   uint32_t TDFV_after = *TDFV;
+   *TLR = p->tot_len;
+   *ISR = (unsigned int) 0xFFFFFFFF;
+   uint32_t TDFV_V = *TDFV;
+   or1k_critical_end(restore_2);
+   return ERR_OK;
+}
+
+/*
+ * netif_status_callback: callback function if netif status changes
+ */
+
+static void netif_status_callback(struct netif *netif)
+{
+   printf("dhcp_supplied_address: %i\n", dhcp_supplied_address(netif));
+   printf("netif status changed %s\n", ip4addr_ntoa(netif_ip4_addr(netif)));
+}
+
+/*
+ * my_init: Initialization function for network interface netif
+ */
+static err_t my_init(struct netif *netif)
+{
+   netif->linkoutput = netif_output;
+   netif->output = etharp_output;
+   netif->output_ip6 = ethip6_output;
+   netif->mtu = ETHERNET_MTU;
+   netif->flags = NETIF_FLAG_BROADCAST | NETIF_FLAG_ETHARP | NETIF_FLAG_ETHERNET
+         | NETIF_FLAG_IGMP | NETIF_FLAG_MLD6;
+   MIB2_INIT_NETIF(netif, snmp_ifType_ethernet_csmacd, 100000000);
+
+   SMEMCPY(netif->hwaddr, MYMACADDRESS, sizeof(netif->hwaddr));
+   netif->hwaddr_len = sizeof(netif->hwaddr);
+   return ERR_OK;
+}
+
+void init()
+{
+
+}
+
+#if LWIP_DHCP
+/*
+ * DHCP Init
+ */
+void dhcp_init(struct netif *netif)
+{
+   err_t error_dhcp;
+   /* Start DHCP and HTTPD */
+   error_dhcp = dhcp_start(netif);
+   if (error_dhcp != ERR_OK) {
+      printf("DHCP Error occurred - out of memory.\n");
+   } else {
+      printf("DHCP started.\n");
+      u8_t myip_dhcp;
+      myip_dhcp = dhcp_supplied_address(netif);
+      printf("main: ip address now: %x\n", myip_dhcp);
+   }
+}
+#endif // LWIP_DHCP
+
+void main(void)
+{
+   app_init();
+   lwip_init();
+   struct netif netif;
+   struct netif *netif_control;
+   err_t err_ipv6_netif;
```

Comment:
Please reverse the order here: first the variable declarations, then the init functions.